### PR TITLE
refactor: Rewrite config parser to respect the order

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR will fix https://github.com/datafuselabs/databend/issues/4362

To make our config parsed in a more understandable way, I built the following two libs:

- [serde-bridge](https://github.com/Xuanwo/serde-bridge): A bridge between different serde implementations.
- [serfig](https://github.com/Xuanwo/serfig): Layered configuration system built upon serde

Now, our config is parsed in the following ways:

- Load from files (powered by toml and so on)
- Load from env (powered by [serde-env](https://github.com/Xuanwo/serde-env/pull/1))
- Load from clap args

All of them are bridged by `serde-bridge`.

With this PR, we will:

- Avoid args conflict via the new style of args naming style
- Remove `env-helper!` as we are loading by [envy](https://github.com/softprops/envy)

Our config will be more **consistent** and **extensible**.

### args naming

- From `access-key-id` to `storage-s3-access-key-id`
- From `region` to `storage-s3-region`
- Remove `load_from_args`, `load_from_env`, `load_from_file` from `Config`

All storage service-related args will be prefixed by `storage-<service>-` instead.

### env naming

- From `S3_STORAGE_ACCESS_KEY_ID` to `STORAGE_S3_ACCESS_KEY_ID`
- Every `env` will map to an arg directly.

## Preview

image

## Current Status

PR is ready for review now!

Works are listed here

- [ ] Make sure all test cases passed
- [x] Merge with main to adapt to new changes
- [x] Make a release for `serde-bridge`, `serfig` and `serde-env`.
- [x] Update all docs related to this change.

## Changelog

- Improvement

## Related Issues

Fixes https://github.com/datafuselabs/databend/issues/4362

### CODE REVIEW

**Comment by ZhiHanZ** [comment]
I prefer load order like:
1. from argument
2. from configfile
3. from environment variable

to treat environment variable as first-class citizen and respect the configuration from configfile, and it would also cause painful breaking change if config from configfile could overlap settings from environment variables

**Comment by Xuanwo** [comment]
> I prefer load order like:
> 
> 1. from argument
> 2. from configfile
> 3. from environment variable
> 
> to treat environment variable as first-class citizen and respect the configuration from configfile, and it would also cause painful breaking change if config from configfile could overlap settings from environment variables

Yep, this proposal does implement this order, next layer will overwrite the same value for the current layer. That means we will take env first, and `overwrite` by `config` and `argument`.

File: query/Cargo.toml [hunk 1]
```diff
@@ -106,6 +106,8 @@ tonic = "=0.6.2"
 typetag = "0.1.8"
 uuid = { version = "0.8.2", features = ["serde", "v4"] }
 walkdir = "2.3.2"
+serfig = {path="../../../xuanwo/serfig"}
+serde-bridge = {path="../../../xuanwo/serde-bridge"}
```
Comment (drmingdrmer) [inline query/Cargo.toml:110]: Can not find the file 🤔 ?

Comment (Xuanwo) [reply query/Cargo.toml:110]: Yep, this PR is a demo to show my idea and is not workable.

I expect to make it works tomorrow :rocket:

Comment (Xuanwo) [reply query/Cargo.toml]: The PR is buildable now!

**Comment by Xuanwo** [comment]
envy is not suitable for databend as it can't deserialize value in nested structs. I build a new lib [serde-env](https://github.com/Xuanwo/serde-env) to fix it.

Now, our PR is buildable. I'm going to move to the next step: make PR ready for review :rocket:

**Comment by Xuanwo** [comment]
image

Oh, I love deleting code (

**Comment by Xuanwo** [comment]
I added a test for override: https://github.com/datafuselabs/databend/pull/4971/files#diff-a8af230cc9f832a19df365f4ddd5cf49cc3bd752705c27445008c928d7f2041dR193-R295

@ZhiHanZ Please help check whether this behavior is expected.

File: docs/doc/10-deploy/11-cluster-k8s-minio.md [hunk 2]
```diff
@@ -129,4 +129,6 @@ SELECT count(*) FROM t1;
 +----------+
 | 10000000 |
 +----------+
-```
\ No newline at end of file
+1 row in set (0.02 sec)
```
Comment (BohuTANG) [inline docs/doc/10-deploy/11-cluster-k8s-minio.md:132]: These two lines should removed:
```
1 row in set (0.02 sec)
Read 1 rows, 1 B in 0.001 sec., 749.34 rows/sec., 749.34 B/sec.
```

**Comment by flaneur2020** [comment]
I guess we can take a versioned config approach like: https://github.com/k3d-io/k3d/tree/main/pkg/config

on introducing incompatibility config changes, we can bounce up the config version with a migration logic, like https://github.com/flaneur2020/k3d/blob/412950bd84bcee0050c42515ef19356689152328/pkg/config/v1alpha3/migrations.go#L41

thus the older configs can keep compatible until the annual deprecation cycle, allowing users to migrate the configs after reading our deprecation news but before the actual api got break.

**Comment by Xuanwo** [comment]
> I guess we can consider a versioned config approach like: https://github.com/k3d-io/k3d/tree/main/pkg/config

I will add a version id first in this PR.

**Comment by Xuanwo** [comment]
`test_stateless_standalone_macos` seems not very stable.

**Comment by Xuanwo** [comment]
ping @BohuTANG @flaneur2020 for another approve.

**Review by BohuTANG — APPROVED** [review]
lgtm

**Pull request merged**